File: packages/agent_security/android/src/main/kotlin/com/chapter2/agentkit/guard/AgentKitSecurityGuard.kt
package com.chapter2.agentkit.guard

import com.chapter2.agentkit.models.AgentKitSecurityException
import com.chapter2.agentkit.models.WorldIdSignerBinding

public sealed class AgentKitDecision {
    data object AutonomousAllow : AgentKitDecision()
    data class EscalateToBiometricHumanApproval(val reason: String) : AgentKitDecision()
}

public class AgentKitSecurityGuard {
    fun evaluateAgentAction(
        amountUnits: ULong,
        maxAutonomousCap: ULong = 100_000_000uL,
        humanBinding: WorldIdSignerBinding?,
        currentTimestamp: Long = System.currentTimeMillis()
    ): Result<AgentKitDecision> {
        if (amountUnits <= maxAutonomousCap) {
            return Result.success(AgentKitDecision.AutonomousAllow)
        }

        if (humanBinding == null) {
            return Result.failure(
                AgentKitSecurityException.AutonomousLimitExceeded(
                    amount = amountUnits,
                    maxAutonomousCap = maxAutonomousCap
                )
            )
        }

        if (!humanBinding.isValid(currentTimestamp)) {
            return Result.failure(
                AgentKitSecurityException.BindingExpired90Days(
                    expiredAtTimestamp = humanBinding.expiresAtTimestamp
                )
            )
        }

        return Result.success(
            AgentKitDecision.EscalateToBiometricHumanApproval(
                "Action exceeds autonomous limit ($amountUnits > $maxAutonomousCap). Verified human operator required."
            )
        )
    }
}

File: packages/agent_security/ios/Classes/Guard/AgentKitSecurityGuard.swift
import Foundation

public enum AgentKitDecision: Equatable, Sendable {
    case autonomousAllow
    case escalateToBiometricHumanApproval(reason: String)
}

public final class AgentKitSecurityGuard: Sendable {
    public init() {}

    public func evaluateAgentAction(
        amountUnits: UInt64,
        maxAutonomousCap: UInt64 = 100_000_000,
        humanBinding: WorldIdSignerBinding?,
        currentDate: Date = Date()
    ) -> Result<AgentKitDecision, AgentKitSecurityError> {
        if amountUnits <= maxAutonomousCap {
            return .success(.autonomousAllow)
        }

        guard let binding = humanBinding else {
            return .failure(.autonomousLimitExceeded(amount: amountUnits, maxAutonomousCap: maxAutonomousCap))
        }

        guard binding.isValid(at: currentDate) else {
            return .failure(.bindingExpired90Days(expiredAt: binding.expiresAt))
        }

        return .success(.escalateToBiometricHumanApproval(
            reason: "Action exceeds autonomous ceiling (\(amountUnits) > \(maxAutonomousCap)). Verified human operator required."
        ))
    }
}
